internal/client/api/leaves.go

```diff
@@ -0,0 +1,351 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package api
+
+import (
+	"slices"
+
+	"github.com/ChainSafe/gossamer/internal/primitives/core/hash"
+	"github.com/ChainSafe/gossamer/internal/primitives/database"
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+	"github.com/tidwall/btree"
+)
+
+type leafSetItem[H comparable, N runtime.Number] struct {
+	hash   H
+	number N
+}
+
+// ImportOutcome privately contains the inserted and removed leaves after an import action.
+type ImportOutcome[H comparable, N runtime.Number] struct {
+	inserted leafSetItem[H, N]
+	removed  *H
+}
+
+// RemoveOutcome privates contains the inserted and removed leaves after a remove action.
+type RemoveOutcome[H comparable, N runtime.Number] struct {
+	inserted *H
+	removed  leafSetItem[H, N]
+}
+
+// FinalizationOutcome contains the removed leaves after a finalization action.
+type FinalizationOutcome[H comparable, N runtime.Number] struct {
+	removed btree.Map[N, []H]
+}
+
+// Leaves returns the leaves that were removed after a finalization action.
+func (fo FinalizationOutcome[H, N]) Leaves() []H {
+	leaves := make([]H, 0)
+	for _, hashes := range fo.removed.Values() {
+		leaves = append(leaves, hashes...)
+	}
+	return leaves
+}
+
+// list of leaf hashes ordered by number (descending).
+// stored in memory for fast access.
+// this allows very fast checking and modification of active leaves.
+type LeafSet[H comparable, N runtime.Number] struct {
+	storage btree.Map[N, []H]
+}
+
+// NewLeafSet is constructor for a new, blank `LeafSet`.
+func NewLeafSet[H comparable, N runtime.Number]() LeafSet[H, N] {
+	return LeafSet[H, N]{
+		storage: *btree.NewMap[N, []H](0),
+	}
+}
+
+// NewLeafSetFromDB will read the leaf list from the DB, using given prefix for keys.
+func NewLeafSetFromDB[H comparable, N runtime.Number](
+	db database.Database[hash.H256], column uint32, prefix []byte,
+) (LeafSet[H, N], error) {
+	storage := btree.NewMap[N, []H](0)
+
+	leaves := db.Get(database.ColumnID(column), prefix)
+	if leaves != nil {
+		type numberHashes struct {
+			Number N
+			Hashes []H
+		}
+		var vals []numberHashes
+		err := scale.Unmarshal(leaves, &vals)
+		if err != nil {
+			return LeafSet[H, N]{}, err
+		}
+		for _, nh := range vals {
+			storage.Set(nh.Number, nh.Hashes)
+		}
+	}
+	return LeafSet[H, N]{
+		storage: *storage,
+	}, nil
+}
+
+// Import will update the leaf list on import.
```

Comment:
Could you improve the comment to explain why importing needs to remove the parent hash?

Reply:
This is a leaf block set, so if a block is appended to a parent, it's no longer a leaf.